File: javascript/packages/linter/src/rules/erb-no-silent-statement.ts
import { BaseRuleVisitor } from "./rule-utils.js"
import { ParserRule } from "../types.js"

import { isERBOutputNode, PrismNode } from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult, ERBContentNode, ParserOptions } from "@herb-tools/core"

function isAssignmentNode(prismNode: PrismNode): boolean {
  const type: string = prismNode?.constructor?.name
  if (!type) return false

  return type.endsWith("WriteNode")
}

class ERBNoSilentStatementVisitor extends BaseRuleVisitor {
  visitERBContentNode(node: ERBContentNode): void {
    if (isERBOutputNode(node)) return

    const prismNode = node.prismNode
    if (!prismNode) return

    if (isAssignmentNode(prismNode)) return

    const content = node.content?.value?.trim()
    if (!content) return

    this.addOffense(
      `Avoid using silent ERB tags for statements. Move \`${content}\` to a controller, helper, or presenter.`,
      node.location,
    )
  }
}

export class ERBNoSilentStatementRule extends ParserRule {
  static ruleName = "erb-no-silent-statement"

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: false,
      severity: "warning"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      prism_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ERBNoSilentStatementVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/erb_no_silent_statement.rs
use crate::utils::erb_utils::is_output_tag_opening;
use crate::utils::prism_utils::{is_assignment_node, is_control_flow_node, is_side_effect_call, unwrap_modifier_statement};

use herb::nodes::ERBContentNode;
use herb::Visitor;

rule_visitor!(ERBNoSilentStatementVisitor);
define_parser_rule!(
  ERBNoSilentStatementRule,
  "erb-no-silent-statement",
  Warning,
  ERBNoSilentStatementVisitor,
  enabled: false,
  parser_options: { prism_nodes: true }
);

impl Visitor for ERBNoSilentStatementVisitor {
  fn visit_erb_content_node(&mut self, node: &ERBContentNode) {
    let tag_opening = node.tag_opening.as_ref().map(|token| token.value.as_str()).unwrap_or("");

    if is_output_tag_opening(tag_opening) {
      return;
    }

    let prism_node = match node.prism_node_ast {
      Some(ref prism_node) => prism_node,
      None => return,
    };

    if is_assignment_node(prism_node) {
      return;
    }

    let statement = unwrap_modifier_statement(prism_node);

    if is_control_flow_node(statement) || is_side_effect_call(statement) {
      return;
    }

    let content = node.content.as_ref().map(|token| token.value.trim()).unwrap_or("");

    if content.is_empty() {
      return;
    }

    self.add_offense(
      format!(
        "Avoid using silent ERB tags for statements. Move `{}` to a controller, helper, or presenter.",
        content
      ),
      node.location.clone(),
    );
  }
}
